src/txmempool.cpp

```diff
@@ -429,5 +439,47 @@ bool CCoinsViewMemPool::HaveCoins(const uint256 &txid) const {
 
 size_t CTxMemPool::DynamicMemoryUsage() const {
     LOCK(cs);
-    return memusage::DynamicUsage(mapTx) + memusage::DynamicUsage(mapNextTx) + memusage::DynamicUsage(mapDeltas) + cachedInnerUsage;
+    // Estimate the overhead of mapTx to be 5 pointers + an allocation, as no exact formula for boost::multi_index_contained is implemented.
+    return memusage::MallocUsage(sizeof(CTxMemPoolEntry) + 5 * sizeof(void*)) * mapTx.size() + memusage::DynamicUsage(mapNextTx) + memusage::DynamicUsage(mapDeltas) + cachedInnerUsage;
+}
+
+size_t CTxMemPool::GuessDynamicMemoryUsage(const CTxMemPoolEntry& entry) const {
+    return memusage::MallocUsage(sizeof(CTxMemPoolEntry) + 5 * sizeof(void*)) + entry.DynamicMemoryUsage() + memusage::IncrementalDynamicUsage(mapNextTx) * entry.GetTx().vin.size();
+}
+
+bool CTxMemPool::StageTrimToSize(size_t sizelimit, const CAmount& maxfeeremove, std::set<uint256>& stage, CAmount& totalfeeremoved, size_t& totalsizeremove) {
+    size_t expsize = DynamicMemoryUsage();
+    indexed_transaction_set::nth_index<1>::type::reverse_iterator it = mapTx.get<1>().rbegin();
+    while (expsize > sizelimit && it != mapTx.get<1>().rend()) {
+        std::deque<uint256> todo;
+        todo.push_back(it->GetTx().GetHash());
+        while (!todo.empty()) {
+            uint256 donow = todo.front();
+            const CTxMemPoolEntry* origTx = &*mapTx.find(donow);
+            todo.pop_front();
+            if (!stage.count(donow)) {
+                stage.insert(donow);
+                totalfeeremoved += origTx->GetFee();
+                if (totalfeeremoved > maxfeeremove) {
+                    return false;
```

Comment:
What happens if the lowest fee-rate transaction in the pool is a parent transaction to a chain of high fee transactions?  Wouldn't that prevent any eviction from happening, even though there may be other transactions that could be removed to make room?